hub/tasks/x_event_source.py

```diff
@@ -0,0 +1,164 @@
+
+from apscheduler.schedulers.asyncio import AsyncIOScheduler
+from apscheduler.triggers.interval import IntervalTrigger
+from nearai.shared.client_config import DEFAULT_MODEL
+
+from hub.api.v1.auth import AuthToken
+from hub.api.v1.thread_routes import RunCreateParamsBase, ThreadModel, _create_thread, create_run
+from hub.tasks.twitter_client import get_latest_mentions
+
+RESET_BLOCK_ID_ON_START = True  # reusing setting name for now
+
+# Path to store the last read tweet id
+TWEET_STATUS_FILE = "last_tweet.txt"
+
+scheduler = AsyncIOScheduler()
+
+# todo pull from registry 'integration' item
+EVENT_SOURCE_CONFIG = {"user_id": "1867270324649160704"}
+
+# todo pull from agent metadata
+LISTENER_CONFIG = {"agent_name": "flatirons.near/nearsecretagent/0.0.1"}
+
+
+# Function to get the last read tweet timestamp from a file
+async def get_last_tweet_timestamp():
+    if os.path.exists(TWEET_STATUS_FILE):
+        with open(TWEET_STATUS_FILE, "r") as f:
+            return int(f.read().strip())
+    return None
+
+
+# Function to save the last blockId to a file
+def save_last_tweet_time(tweet_timestamp):
+    print("Saving last tweet time", tweet_timestamp)
+    with open(TWEET_STATUS_FILE, "w") as f:
+        f.write(str(tweet_timestamp))
+
+
+def user_rate_limit_exceeded(tweet):
+    # todo implement
+    pass
+
+
+def rate_limit_reply(tweet):
+    # if the user has received a rate limit reply already today, do nothing.
+    # response = "So many questions, try again tomorrow."  # pull from agent metadata?
+    # todo implement
+    pass
+
+
+async def listener_function(auth_token, tweet):
+    message = tweet.text
+    sender = tweet.id
+
+    # if user_rate_limit_exceeded(tweet):
+    #     rate_limit_reply(tweet)
+    #     return
+
+    # todo one thread per twitter sender
+
+    await run_agent(
+        LISTENER_CONFIG["agent_name"],
+        message,
+        sender,
+        {},
+        auth_token,
+    )
+
+
+# @todo change to hub authentication
+def load_auth_token():
+    from nearai.config import Config, load_config_file
+
+    app_config = Config()
+    # Update config from global config file
+    config_data = load_config_file(local=False)
+    app_config = app_config.update_with(config_data)
+
+    return app_config.auth
+
+
+async def run_agent(agent, message, sender, data, auth_token: AuthToken):
+    if not (agent and message):
+        return print("Missing data")
+
+    thread_model = ThreadModel(
+        meta_data={
+            "agent_ids": [agent],
+        },
+        tool_resources=None,
+        owner_id=auth_token.account_id,
+    )
+
+    thread = _create_thread(thread_model, auth=auth_token)
+
+    run_params = RunCreateParamsBase(
+        assistant_id=agent,
+        model=DEFAULT_MODEL,
+        instructions=None,
+        tools=None,
+        metadata=None,
+        include=[],
+        additional_instructions=None,
+        additional_messages=[{"content": message, "role": "user"}],
```

Comment:
It might be help helpful to provide tweet_id too, so agent will read extra context itself if needed.
so "content": json.dumps({"message": ...,  "tweet_id": ..})

Reply:
Yeah, was working on this data format in the agent. Also will remove this function and use the one in nearevents